File: src/lib/tama/characters.ts
import type { AdultId, CharacterId, CharacterStats } from "./types";

export const CHARACTERS: Record<CharacterId, CharacterStats> = {
  egg: {
    id: "egg",
    name: "Egg",
    short: "Egg",
    stage: "egg",
    wakeHour: null,
    sleepHour: null,
    hungryLossMin: 999,
    happyLossMin: 999,
    sicknessMin: 9999,
    shots: 0,
    minWeight: 0,
    maxWeight: 0,
    evoMin: 5,
    disciplineCountdown: null,
    initialDiscipline: 0,
    bites: 0,
    gameWinPct: 0,
    lifespan: "5 min",
    blurb: "Set the clock. The egg hatches about five minutes later. No care yet.",
  },
  babytchi: {
    id: "babytchi",
    name: "Babytchi",
    short: "Baby",
    stage: "baby",
    wakeHour: null,
    sleepHour: null,
    hungryLossMin: 3,
    happyLossMin: 4,
    sicknessMin: 45,
    shots: 2,
    minWeight: 5,
    maxWeight: 5,
    evoMin: 65,
    disciplineCountdown: null,
    initialDiscipline: 0,
    bites: 4,
    gameWinPct: 50,
    lifespan: "65 min",
    blurb: "Always becomes Marutchi. Hearts drain every 3–4 minutes. First poop around 15 minutes, a nap around 40. Care here does not change the next form.",
  },
  marutchi: {
    id: "marutchi",
    name: "Marutchi",
    short: "Child",
    stage: "child",
    wakeHour: 9,
    sleepHour: 20,
    hungryLossMin: 50,
    happyLossMin: 60,
    sicknessMin: 990,
    shots: 2,
    minWeight: 10,
    maxWeight: 99,
    evoMin: 2880,
    disciplineCountdown: 6,
    initialDiscipline: 0,
    bites: 4,
    gameWinPct: 50,
    lifespan: "48 hours",
    blurb: "The fork. About 48 hours in this form. Under 3 care mistakes keeps the good teen (Tamatchi). 3 or more becomes Kuchitamatchi and locks out Mametchi.",
  },
  tamatchi: {
    id: "tamatchi",
    name: "Tamatchi",
    short: "Teen",
    stage: "teen",
    wakeHour: 9,
    sleepHour: 21,
    hungryLossMin: 75,
    happyLossMin: 85,
    sicknessMin: 1656,
    shots: 2,
    minWeight: 20,
    maxWeight: 99,
    evoMin: 4320,
    disciplineCountdown: 6,
    initialDiscipline: 0,
    bites: 2,
    gameWinPct: 50,
    lifespan: "72 hours",
    blurb: "Good teen. About 72 hours in this form. Can become any adult. Type 1 (under 3 discipline mistakes as Marutchi) is required for Mametchi.",
  },
  kuchitamatchi: {
    id: "kuchitamatchi",
    name: "Kuchitamatchi",
    short: "Teen",
    stage: "teen",
    wakeHour: 9,
    sleepHour: 21,
    hungryLossMin: 75,
    happyLossMin: 85,
    sicknessMin: 660,
    shots: 2,
    minWeight: 20,
    maxWeight: 99,
    evoMin: 2880,
    disciplineCountdown: 6,
    initialDiscipline: 0,
    bites: 4,
    gameWinPct: 50,
    lifespan: "48 hours",
    blurb: "Poor-care teen. About 48 hours in this form. Only Kuchipatchi, Nyorotchi, or Tarakotchi from here. Care mistakes no longer change the adult — discipline mistakes do.",
  },
  mametchi: {
    id: "mametchi",
    name: "Mametchi",
    short: "Adult",
    stage: "adult",
    wakeHour: 9,
    sleepHour: 22,
    hungryLossMin: 81,
    happyLossMin: 91,
    sicknessMin: 3900,
    shots: 1,
    minWeight: 30,
    maxWeight: 99,
    evoMin: 4095,
    disciplineCountdown: null,
    initialDiscipline: 100,
    bites: 2,
    gameWinPct: 50,
    lifespan: "15–16 days",
    blurb: "The prize adult. Needs Tamatchi type 1, under 3 care mistakes total, and zero discipline mistakes.",
  },
  ginjirotchi: {
    id: "ginjirotchi",
    name: "Ginjirotchi",
    short: "Adult",
    stage: "adult",
    wakeHour: 9,
    sleepHour: 22,
    hungryLossMin: 81,
    happyLossMin: 91,
    sicknessMin: 2808,
    shots: 1,
    minWeight: 30,
    maxWeight: 99,
    evoMin: 3120,
    disciplineCountdown: 7,
    initialDiscipline: 50,
    bites: 2,
    gameWinPct: 50,
    lifespan: "11–12 days",
    blurb: "Good-care adult from Tamatchi with exactly one discipline mistake (type 1) or low-discipline type 2.",
  },
  maskutchi: {
    id: "maskutchi",
    name: "Maskutchi",
    short: "Adult",
    stage: "adult",
    wakeHour: 11,
    sleepHour: 23,
    hungryLossMin: 55,
    happyLossMin: 65,
    sicknessMin: 2592,
    shots: 1,
    minWeight: 30,
    maxWeight: 99,
    evoMin: 2880,
    disciplineCountdown: 7,
    initialDiscipline: 0,
    bites: 2,
    gameWinPct: 31,
    lifespan: "15–16 days",
    blurb: "Sleeps late (11pm–11am). Hearts drain faster than other good adults. Type 2 Maskutchi can become Bill / Oyajitchi.",
  },
  kuchipatchi: {
    id: "kuchipatchi",
    name: "Kuchipatchi",
    short: "Adult",
    stage: "adult",
    wakeHour: 9,
    sleepHour: 22,
    hungryLossMin: 60,
    happyLossMin: 70,
    sicknessMin: 1170,
    shots: 2,
    minWeight: 20,
    maxWeight: 99,
    evoMin: 1560,
    disciplineCountdown: null,
    initialDiscipline: 100,
    bites: 2,
    gameWinPct: 69,
    lifespan: "5–6 days",
    blurb: "Chubby happy adult. Needs some care mistakes and few discipline mistakes. Easiest “healthy” poor-care result.",
  },
  nyorotchi: {
    id: "nyorotchi",
    name: "Nyorotchi",
    short: "Adult",
    stage: "adult",
    wakeHour: 9,
    sleepHour: 22,
    hungryLossMin: 60,
    happyLossMin: 70,
    sicknessMin: 360,
    shots: 3,
    minWeight: 10,
    maxWeight: 99,
    evoMin: 780,
    disciplineCountdown: 7,
    initialDiscipline: 50,
    bites: 4,
    gameWinPct: 50,
    lifespan: "2–3 days",
    blurb: "Gets sick easily (needs 3 shots). Short life. Comes from mixed discipline mistakes on a poor-care path.",
  },
  tarakotchi: {
    id: "tarakotchi",
    name: "Tarakotchi",
    short: "Adult",
    stage: "adult",
    wakeHour: 10,
    sleepHour: 22,
    hungryLossMin: 45,
    happyLossMin: 50,
    sicknessMin: 660,
    shots: 2,
    minWeight: 20,
    maxWeight: 99,
    evoMin: 1440,
    disciplineCountdown: 7,
    initialDiscipline: 0,
    bites: 2,
    gameWinPct: 50,
    lifespan: "3–4 days",
    blurb: "The neglect adult. Fast heart drain, wakes at 10am. Lots of discipline mistakes.",
  },
  oyajitchi: {
    id: "oyajitchi",
    name: "Oyajitchi",
    short: "Secret",
    stage: "secret",
    wakeHour: 9,
    sleepHour: 22,
    hungryLossMin: 81,
    happyLossMin: 91,
    sicknessMin: 3900,
    shots: 1,
    minWeight: 30,
    maxWeight: 99,
    evoMin: 4095,
    disciplineCountdown: null,
    initialDiscipline: 100,
    bites: 2,
    gameWinPct: 50,
    lifespan: "15–16 days",
    blurb: "Japanese secret. Raise a type-2 Tamatchi into Maskutchi (never scold), then keep it alive about 4 more days.",
  },
  bill: {
    id: "bill",
    name: "Bill",
    short: "Secret",
    stage: "secret",
    wakeHour: 9,
    sleepHour: 22,
    hungryLossMin: 81,
    happyLossMin: 91,
    sicknessMin: 3900,
    shots: 1,
    minWeight: 30,
    maxWeight: 99,
    evoMin: 4095,
    disciplineCountdown: null,
    initialDiscipline: 100,
    bites: 2,
    gameWinPct: 50,
    lifespan: "15–16 days",
    blurb: "English secret (same path as Oyajitchi). Type-2 Maskutchi, then wait. Care after Maskutchi only has to keep it alive.",
  },
};

export const ADULT_IDS: AdultId[] = [
  "mametchi",
  "ginjirotchi",
  "maskutchi",
  "kuchipatchi",
  "nyorotchi",
  "tarakotchi",
  "oyajitchi",
  "bill",
];

export const STAGE_ORDER: CharacterId[] = [
  "egg",
  "babytchi",
  "marutchi",
  "tamatchi",
  "kuchitamatchi",
  "mametchi",
  "ginjirotchi",
  "maskutchi",
  "kuchipatchi",
  "nyorotchi",
  "tarakotchi",
  "oyajitchi",
  "bill",
];

export const CARE_WINDOW_MS = 15 * 60 * 1000;
export const EGG_MS = 5 * 60 * 1000;
/** Observed P1 stage lengths (wall clock). The ROM “base time” field is a different unit and was evolving Marutchi a day early. */
export const BABY_MS = 65 * 60 * 1000;
export const MARU_MS = 48 * 60 * 60 * 1000;
export const TAMATCHI_MS = 72 * 60 * 60 * 1000;
export const KUCHITA_MS = 48 * 60 * 60 * 1000;
export const SECRET_WAIT_MS = 4 * 24 * 60 * 60 * 1000;
export const BABY_POOP_MS = 15 * 60 * 1000;
export const BABY_NAP_MS = 40 * 60 * 1000;

export const POOP_INTERVAL_MIN: Record<CharacterId, number> = {
  egg: 9999,
  babytchi: 15,
  marutchi: 90,
  tamatchi: 120,
  kuchitamatchi: 100,
  mametchi: 180,
  ginjirotchi: 180,
  maskutchi: 150,
  kuchipatchi: 150,
  nyorotchi: 90,
  tarakotchi: 90,
  oyajitchi: 180,
  bill: 180,
};

export function statsFor(id: CharacterId): CharacterStats {
  return CHARACTERS[id];
}

export function displayName(id: CharacterId): string {
  return CHARACTERS[id].name;
}

export function isAdultForm(id: CharacterId): boolean {
  return CHARACTERS[id].stage === "adult" || CHARACTERS[id].stage === "secret";
}

export function secretForRegion(region: "en" | "jp"): AdultId {
  return region === "jp" ? "oyajitchi" : "bill";
}

/** Original P1 floors once age has eaten the base interval. */
export const HUNGER_LOSS_FLOOR_MIN = 7;
export const HAPPY_LOSS_FLOOR_MIN = 9;

/** Base rate minus 1 minute per year of age. Baby stays at 3/4. */
export function hungerLossMin(form: CharacterId, age: number): number {
  const base = CHARACTERS[form].hungryLossMin;
  if (form === "egg" || form === "babytchi") return base;
  return Math.max(HUNGER_LOSS_FLOOR_MIN, base - Math.max(0, age));
}

export function happyLossMin(form: CharacterId, age: number): number {
  const base = CHARACTERS[form].happyLossMin;
  if (form === "egg" || form === "babytchi") return base;
  return Math.max(HAPPY_LOSS_FLOOR_MIN, base - Math.max(0, age));
}

File: android/app/src/main/java/com/hatchwatch/app/engine/Characters.kt
package com.hatchwatch.app.engine

object Characters {
    const val CARE_WINDOW_MS = 15L * 60 * 1000
    const val EGG_MS = 5L * 60 * 1000
    const val BABY_MS = 65L * 60 * 1000
    const val MARU_MS = 48L * 60 * 60 * 1000
    const val TAMATCHI_MS = 72L * 60 * 60 * 1000
    const val KUCHITA_MS = 48L * 60 * 60 * 1000
    const val SECRET_WAIT_MS = 4L * 24 * 60 * 60 * 1000
    const val BABY_POOP_MS = 15L * 60 * 1000
    const val BABY_NAP_MS = 40L * 60 * 1000
    const val HUNGER_FLOOR = 7
    const val HAPPY_FLOOR = 9
    const val WARN_LEAD_MS = 2L * 60 * 1000

    val ADULT_IDS = listOf(
        CharacterId.mametchi, CharacterId.ginjirotchi, CharacterId.maskutchi,
        CharacterId.kuchipatchi, CharacterId.nyorotchi, CharacterId.tarakotchi,
        CharacterId.oyajitchi, CharacterId.bill,
    )

    val STAGE_ORDER = CharacterId.entries

    private fun c(
        id: CharacterId, name: String, short: String, stage: Stage,
        wake: Int?, sleep: Int?, hungry: Int, happy: Int, sick: Int,
        shots: Int, minW: Int, maxW: Int, evo: Int, disc: Int?,
        initDisc: Int, bites: Int, win: Int, life: String, blurb: String,
    ) = CharacterStats(
        id, name, short, stage, wake, sleep, hungry, happy, sick,
        shots, minW, maxW, evo, disc, initDisc, bites, win, life, blurb,
    )

    val ALL: Map<CharacterId, CharacterStats> = listOf(
        c(CharacterId.egg, "Egg", "Egg", Stage.egg, null, null, 999, 999, 9999, 0, 0, 0, 5, null, 0, 0, 0, "5 min", "Set the clock. The egg hatches about five minutes later. No care yet."),
        c(CharacterId.babytchi, "Babytchi", "Baby", Stage.baby, null, null, 3, 4, 33, 2, 5, 5, 65, null, 0, 4, 50, "65 min", "Always becomes Marutchi. Hearts drain every 3–4 minutes. First poop at 15 minutes, again at 45. Skull around 33 minutes. Nap around 40."),
        c(CharacterId.marutchi, "Marutchi", "Child", Stage.child, 9, 20, 50, 60, 990, 2, 10, 99, 2880, 6, 0, 4, 50, "48 hours", "The fork. About 48 hours. Under 3 care mistakes keeps Tamatchi. 3 or more becomes Kuchitamatchi."),
        c(CharacterId.tamatchi, "Tamatchi", "Teen", Stage.teen, 9, 21, 75, 85, 1656, 2, 20, 99, 4320, 6, 0, 2, 50, "72 hours", "Good teen. About 72 hours. Type 1 is required for Mametchi."),
        c(CharacterId.kuchitamatchi, "Kuchitamatchi", "Teen", Stage.teen, 9, 21, 75, 85, 660, 2, 20, 99, 2880, 6, 0, 4, 50, "48 hours", "Poor-care teen. About 48 hours. Only Kuchipatchi, Nyorotchi, or Tarakotchi."),
        c(CharacterId.mametchi, "Mametchi", "Adult", Stage.adult, 9, 22, 81, 91, 3900, 1, 30, 99, 4095, null, 100, 2, 50, "15–16 days", "The prize adult. Tamatchi type 1, under 3 care, zero discipline mistakes."),
        c(CharacterId.ginjirotchi, "Ginjirotchi", "Adult", Stage.adult, 9, 22, 81, 91, 2808, 1, 30, 99, 3120, 7, 50, 2, 50, "11–12 days", "Good-care adult from Tamatchi with exactly one discipline mistake."),
        c(CharacterId.maskutchi, "Maskutchi", "Adult", Stage.adult, 11, 23, 55, 65, 2592, 1, 30, 99, 2880, 7, 0, 2, 31, "15–16 days", "Sleeps 11pm–11am. Type 2 can become Bill / Oyajitchi."),
        c(CharacterId.kuchipatchi, "Kuchipatchi", "Adult", Stage.adult, 9, 22, 60, 70, 1170, 2, 20, 99, 1560, null, 100, 2, 69, "5–6 days", "Chubby happy adult. Some care mistakes, few discipline mistakes."),
        c(CharacterId.nyorotchi, "Nyorotchi", "Adult", Stage.adult, 9, 22, 60, 70, 360, 3, 10, 99, 780, 7, 50, 4, 50, "2–3 days", "Gets sick easily. 3 shots. Short life."),
        c(CharacterId.tarakotchi, "Tarakotchi", "Adult", Stage.adult, 10, 22, 45, 50, 660, 2, 20, 99, 1440, 7, 0, 2, 50, "3–4 days", "The neglect adult. Fast heart drain, wakes at 10am."),
        c(CharacterId.oyajitchi, "Oyajitchi", "Secret", Stage.secret, 9, 22, 81, 91, 3900, 1, 30, 99, 4095, null, 100, 2, 50, "15–16 days", "Japanese secret. Type-2 Tamatchi into Maskutchi, then wait ~4 days."),
        c(CharacterId.bill, "Bill", "Secret", Stage.secret, 9, 22, 81, 91, 3900, 1, 30, 99, 4095, null, 100, 2, 50, "15–16 days", "English secret. Same path as Oyajitchi."),
    ).associateBy { it.id }

    fun stats(id: CharacterId) = ALL.getValue(id)
    fun name(id: CharacterId) = stats(id).name
    fun secretForRegion(region: String) = if (region == "jp") CharacterId.oyajitchi else CharacterId.bill

    val POOP_INTERVAL_MIN = mapOf(
        CharacterId.egg to 9999,
        CharacterId.babytchi to 30,
        CharacterId.marutchi to 90,
        CharacterId.tamatchi to 120,
        CharacterId.kuchitamatchi to 100,
        CharacterId.mametchi to 180,
        CharacterId.ginjirotchi to 180,
        CharacterId.maskutchi to 150,
        CharacterId.kuchipatchi to 150,
        CharacterId.nyorotchi to 90,
        CharacterId.tarakotchi to 90,
        CharacterId.oyajitchi to 180,
        CharacterId.bill to 180,
    )

    fun hungerLossMin(form: CharacterId, age: Int): Int {
        val base = stats(form).hungryLossMin
        if (form == CharacterId.egg || form == CharacterId.babytchi) return base
        return maxOf(HUNGER_FLOOR, base - maxOf(0, age))
    }

    fun happyLossMin(form: CharacterId, age: Int): Int {
        val base = stats(form).happyLossMin
        if (form == CharacterId.egg || form == CharacterId.babytchi) return base
        return maxOf(HAPPY_FLOOR, base - maxOf(0, age))
    }
}
